File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int i, j, ans = 0;
        int n = s.size();
        vector<int> mp(128, 0);
        i = j = 0;

        while(i < n) {
            mp[s[i]]++;
            while(mp[s[i]] > 1) {
                mp[s[j]]--;
                j++;
            }
            ans = max(i - j + 1, ans);
            ++i;
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        left, right = 0, 0
        n = len(s)
        if n <= 2: return len(set(s))
        ans  = 0
        mp = {}
        
        while(right < n):
            curChar = s[right]
            
            if curChar in mp:
                ans = max(ans, right - left)
                left = max(mp[curChar] + 1, left)
                del mp[curChar]
            mp[curChar] = right
            right +=1
            
        return max(ans, right - left)
            
            